app/services/course/material/download_service.rb

```diff
@@ -0,0 +1,64 @@
+class Course::Material::DownloadService
+  require 'zip'
+  require 'uri'
+  require 'open-uri'
+
+  class << self
+    # Download the contents and zip them.
+    #
+    # @param [Array<Course::Material|Course::Material::Folder>] contents The contents to be
+    #   downloaded.
+    # @return [String] The url to the zip file.
+    def download_and_zip(contents)
```

Comment:
Ah, I get you. This sounds like a good idea !  

And I think no need to specify the Root folder(it will make the function looks cleaner) ?  
The root folder is just the lowest common ancestor of all the leaves(materials).